network_test.go

```diff
@@ -0,0 +1,392 @@
+}
+
+func TestNetworkInterfacesValidationFails_CNIWithMultiple(t *testing.T) {
+	err := NetworkInterfaces([]NetworkInterface{
+		validCNIInterface,
+		NetworkInterface{
+			StaticConfiguration: &StaticNetworkConfiguration{
+				MacAddress:  "11:22:33:44:55:66",
+				HostDevName: "tap1",
+			},
+		},
+	}).validate(kernelArgsNoIP)
+	assert.Error(t, err, "network interface list with multiple interfaces and CNI configuration should return validation error")
+}
+
+func TestNetworkInterfacesValidationFails_CNIWithKernelArg(t *testing.T) {
+	err := NetworkInterfaces([]NetworkInterface{
+		validCNIInterface,
+	}).validate(kernelArgsWithIP)
+	assert.Error(t, err, "network interface list with CNI config and IP kernel arg should return validation error")
+}
+
+func TestNetworkInterfacesValidationFails_NeitherSpecified(t *testing.T) {
+	err := NetworkInterfaces([]NetworkInterface{{}}).validate(kernelArgsNoIP)
+	assert.Error(t, err, "invalid network config with neither static nor cni configuration did not result in validation error")
+}
+
+func TestNetworkInterfacesValidationFails_BothSpecified(t *testing.T) {
+	err := NetworkInterfaces([]NetworkInterface{{
+		StaticConfiguration: &StaticNetworkConfiguration{
+			MacAddress:  mockMacAddrString,
+			HostDevName: tapName,
+		},
+		CNIConfiguration: validCNIInterface.CNIConfiguration,
+	}}).validate(kernelArgsNoIP)
+	assert.Error(t, err, "invalid network config with both static and cni configuration did not result in validation error")
+}
+
+func TestNetworkMachineCNI(t *testing.T) {
+	if testing.Short() {
+		t.Skip()
+	}
+	fctesting.RequiresRoot(t)
+
+	cniBinPath := []string{"/opt/cni/bin", testDataBin}
+
+	testCNIDir := filepath.Join(testDataPath, "TestCNI")
+	os.RemoveAll(testCNIDir)
+	defer os.RemoveAll(testCNIDir)
+
+	ipamDataDir := filepath.Join(testCNIDir, "ipam.data")
+
+	cniCacheDir := filepath.Join(testCNIDir, "cni.cache")
+	require.NoError(t,
+		os.MkdirAll(cniCacheDir, 0755),
+		"failed to create cni cache dir")
+
+	cniConfDir := filepath.Join(testCNIDir, "cni.conf")
+	require.NoError(t,
+		os.MkdirAll(cniConfDir, 0755),
+		"failed to create cni conf dir")
+
+	const ifName = "veth0"
+	const networkName = "fcnet"
+
+	cniConf := fmt.Sprintf(`{
+  "cniVersion": "0.3.1",
+  "name": "%s",
+  "plugins": [
+    {
+      "type": "ptp",
+      "ipam": {
+        "type": "host-local",
+        "subnet": "192.168.127.0/24",
+        "resolvConf": "/etc/resolv.conf",
+        "dataDir": "%s"
+      }
+    },
+    {
+      "type": "tc-redirect-tap"
+    }
+  ]
+}`, networkName, ipamDataDir)
+
+	cniConfPath := filepath.Join(cniConfDir, fmt.Sprintf("%s.conflist", networkName))
+	require.NoError(t,
+		ioutil.WriteFile(cniConfPath, []byte(cniConf), 0644),
+		"failed to write cni conf file")
+
+	numVMs := 10
+	vmIPs := make(chan string, numVMs)
+
+	var vmWg sync.WaitGroup
+	for i := 0; i < numVMs; i++ {
+		vmWg.Add(1)
+		go func(vmID string) {
```

Comment:
Extracting as a function?

Reply:
Overall this function is very specific to this test. However, I will try to extract some individual parts of it that could be generalized and re-used, which I agree would be good.

Reply:
Yeah, I agree that the function on the goroutine won't be used by others, but just thought that the whole function is long.

Reply:
For sure, I extracted a few decent size chunks out now, let me know what you think.